### ISSUE TITLE
[Support]Emulating `xCAT 2.9.1` ddns behaviour with `xCAT 2.13.2`

### ISSUE THREAD

**Issue by samveen** [xcat2/xcat-core#2974]
**Problem description:**
I've recently upgraded our xCAT installation from 2.9.1 to 2.13.2. After the upgrade, I am facing an issue with ddns updates of aliased hostnames and nic-interface based hostname aliases.

My relevant site configuration is as below:
```
[V2 samveen@xcatmn1 ~]$ ip -4 -o a |awk '{print $2,$4}'
lo 127.0.0.1/8
ens1f0 10.0.16.12/24
eno1 10.2.16.10/20
[V2 samveen@xcatmn1 ~]$  lsdef -t site clustersite |egrep -i 'name|dns'
Object name: clustersite
    dnshandler=ddns
    externaldns=1
    nameservers=10.2.16.10
```

Network configuration:
```
[V2 samveen@xcatmn1 ~]$ lsdef -t network APPS,SRVC 
Object name: SRVC
    dynamicrange=10.2.28.2-10.2.31.253
    mask=255.255.240.0
    mgtifname=eno1
    net=10.2.16.0
    tftpserver=10.2.16.10
Object name: APPS
    dhcpserver=10.0.16.12
    logservers=10.0.16.12
    mask=255.255.248.0
    mgtifname=!remote!ens1f0
    nameservers=10.2.16.11,10.2.16.10
    net=10.4.8.0
    ntpservers=ntp1.infra
    vlanid=101
```
Node configuration:
```
[V2 samveen@xcatmn1 ~]$ lsdef staticorigin2 -i nicips,ip,mac,nichostnamesuffixes
Object name: staticorigin2
    ip=10.2.17.49
    mac=1c:98:ec:1d:32:d0!staticorigin2-priv|14:02:ec:77:c9:d0!staticorigin2-pub
    nichostnamesuffixes.eno49=-pub
    nichostnamesuffixes.eno1=-priv
    nicips.eno1=10.2.17.49
    nicips.eno49=10.4.8.48
```
Post update `makedns` output:
```
[V2 samveen@xcatmn1 ~]$ makedns staticorigin2
Handling staticorigin2-pub in /etc/hosts.
Ignoring host staticorigin2-pub, it does not belong to any nets defined in networks table or the net it belongs to is configured to use an external nameserver.
Handling staticorigin2 in /etc/hosts.
Getting reverse zones, this may take several minutes for a large cluster.
Completed getting reverse zones.
Updating DNS records, this may take several minutes for a large cluster.
Completed updating DNS records.
```

Running `nslookup` on the expected dns entries for the host:
```
[V2 samveen@xcatmn1 ~]$ nslookup staticorigin2
Server:         10.2.16.10
Address:        10.2.16.10#53

Name:   staticorigin2.vndc
Address: 10.2.17.49

[V2 samveen@xcatmn1 ~]$ nslookup staticorigin2-pub
Server:         10.2.16.10
Address:        10.2.16.10#53

** server can't find staticorigin2-priv: NXDOMAIN

[V2 samveen@xcatmn1 ~]$ nslookup staticorigin2-priv
Server:         10.2.16.10
Address:        10.2.16.10#53

** server can't find staticorigin2-priv: NXDOMAIN

```

Expected output (This is ddns records created by `makeds` for `xCAT-2.9.1` for another server `staticorigin1`):
```
[V2 samveen@xcatmn1 ~]$ nslookup staticorigin1
Server:         10.2.16.10
Address:        10.2.16.10#53

Name:   staticorigin1.vndc
Address: 10.2.16.77

[V2 samveen@xcatmn1 ~]$ nslookup staticorigin1-pub
Server:         10.2.16.10
Address:        10.2.16.10#53

Name:   staticorigin1-pub.vndc
Address: 10.4.8.49

[V2 samveen@xcatmn1 ~]$ nslookup staticorigin1-priv
Server:         10.2.16.10
Address:        10.2.16.10#53

staticorigin1-priv.vndc canonical name = staticorigin1.vndc.
Name:   staticorigin1.vndc
Address: 10.2.16.77

```

How do I get the expected output in xCAT 2.13.2?
My challenges are the following:
- xCAT 2.13.2 isn't creating aliased names for the nodename and primary IP (`-priv` alias)
- DNS information for DHCP records in the `network` objects is now being used as the ddns update point: I can either remove it and have bad DNS entries in node `resolv.conf`, **_or_** keep it in place and not get the interface alias records added to ddns.

**Comment by bybai** [comment]
Hi @samveen ， if the staticorigin1-pub ip addr is not in network APPS, will have this error.   Is the staticorigin1-pub defined in /etc/hosts?   Do you execute `makedns` directly?

**Comment by samveen** [comment]
@bybai `staticorigin1-pub` is another server(provided for contrast), which has correct ddns records, (they were created before the upgrade). Now I'm trying to `makedns staticorigin2` and I have the failure.

**Comment by samveen** [comment]
@bybai Can you confirm the `ip,nicips,mac,nichostnamesuffixes` match?
To clarify, I have 2 interfaces: primary interface (`eno1` has nodename+`-priv` alias (also `ip` and `nicips.eno1` are identical), secondary interface `eno49` has `-pub` alias.

Also can I have a look at dns config generated for the subnet (which may give me a clue as well)?

**Comment by samveen** [comment]
@bybai As per xcat man page `man 7 network`
```
 nameservers (networks.nameservers)
             A comma delimited list of DNS servers that each node in this network should use. This value will end up in the nameserver settings of the /etc/resolv.conf on each node in this network.
             If this attribute value is set to the IP address of an xCAT node, make sure DNS is running on it. In a hierarchical cluster, you can also set this attribute to "<xcatmaster>" to mean
             the DNS server for each node in this network should be the node that is managing it (either its service node or the management node).  Used in creating the DHCP network definition, and
             DNS configuration.
```

Is this attribute now used by the `ddns` plugin as the target server to update? In xCAT 2.9.1, it was not. That is the only difference I can think of. Please confirm.

**Comment by bybai** [comment]
@samveen , networks.nameservers is used by `ddns` plugin.

**Comment by bybai** [comment]
Hi @samveen ,
My example:
[root@byrh13 /]# ip -4 -o a |awk '{print $2,$4}'
lo 127.0.0.1/8
eth0 20.4.41.13/8
eth1 10.4.41.13/8

[root@byrh13 /]# lsdef -t site clustersite |egrep -i 'name|dns'
Object name: clustersite
    dnshandler=ddns
    externaldns=1
    nameservers=20.4.41.13

[root@byrh13 /]# lsdef -t network APS,SRVC
Object name: APS
    mask=255.255.248.0
    mtu=1400
    nameservers=20.4.41.4,20.4.41.13
    net=10.4.8.0
Object name: SRVC
    gateway=20.4.41.4
    mask=255.0.0.0
    mgtifname=eth0
    net=20.0.0.0
    tftpserver=20.4.41.13

[root@byrh13 /]# lsdef staticorigin2 -i nicips,ip,mac,nichostnamesuffixes
Object name: staticorigin2
    ip=20.4.41.23
    mac=1c:98:ec:1d:32:d0!staticorigin2-priv|14:02:ec:77:c9:d0!staticorigin2-pub
    nichostnamesuffixes.eno1=-priv
    nichostnamesuffixes.eno49=-pub
    nicips.eno1=20.4.41.23
    nicips.eno49=10.4.8.48
[root@byrh13 /]# cat /etc/hosts|grep staticorigin2
20.4.41.23 staticorigin2 staticorigin2.cluster.com
10.4.8.48 staticorigin2-pub staticorigin2-pub.cluster.com

[root@byrh13 /]# makedns staticorigin2
Handling staticorigin2-pub in /etc/hosts.
Handling staticorigin2 in /etc/hosts.
Getting reverse zones, this may take several minutes for a large cluster.
Completed getting reverse zones.
Updating DNS records, this may take several minutes for a large cluster.
Completed updating DNS records.

[root@byrh13 /]# cat /etc/named.conf
#generated by xCAT: /opt/xcat/sbin/makedns command
options {
        directory "/var/named/";
        allow-recursion { any; };
};

key xcat_key {
        algorithm hmac-md5;
        secret "c0dlYTdYVlc4VVUzNWJTOVV2U1RXckpUOVJVbGZKcEk=";
};

zone "11.4.10.IN-ADDR.ARPA." in {
        type master;
        allow-update {
                key xcat_key;
        };
        file "db.10.4.11";
};
zone "12.4.10.IN-ADDR.ARPA." in {
        type master;
        allow-update {
                key xcat_key;
        };
        file "db.10.4.12";
};
zone "10.4.10.IN-ADDR.ARPA." in {
        type master;
        allow-update {
                key xcat_key;
        };
        file "db.10.4.10";
};
zone "9.4.10.IN-ADDR.ARPA." in {
        type master;
        allow-update {
                key xcat_key;
        };
        file "db.10.4.9";
};
zone "14.4.10.IN-ADDR.ARPA." in {
        type master;
        allow-update {
                key xcat_key;
        };
        file "db.10.4.14";
};
zone "cluster.com" in {
        type master;
        allow-update {
                key xcat_key;
        };
        file "db.cluster.com";
};
zone "15.4.10.IN-ADDR.ARPA." in {
        type master;
        allow-update {
                key xcat_key;
        };
        file "db.10.4.15";
};
zone "20.IN-ADDR.ARPA." in {
        type master;
        allow-update {
                key xcat_key;
        };
        file "db.20";
};
zone "13.4.10.IN-ADDR.ARPA." in {
        type master;
        allow-update {
                key xcat_key;
        };
        file "db.10.4.13";
};
zone "8.4.10.IN-ADDR.ARPA." in {
        type master;
        allow-update {
                key xcat_key;
        };
        file "db.10.4.8";
};

[root@byrh13 /]# ls /var/named/
data        db.10.4.12  db.10.4.15     db.10.4.9  db.cluster.com      named.ca         named.loopback
db.10.4.10  db.10.4.13  db.10.4.8      db.20      db.cluster.com.jnl  named.empty      slaves
db.10.4.11  db.10.4.14  db.10.4.8.jnl  db.20.jnl  dynamic             named.localhost

**Comment by bybai** [comment]
Hi @samveen 
My xcat version:
[root@byrh13 named]# lsxcatd -v
Version 2.13.4 (git commit 493ff9a38cadeece99ea1ae3066572ff15d2b92f, built Fri May  5 01:46:06 EDT 2017)

**Comment by samveen** [comment]
@bybai This is more of a support request and usage patterns rather than a bug. Best to remove the milestone and add the `support` label

**Comment by bybai** [comment]
Hi @samveen ， I added this into support.

**Comment by bybai** [comment]
Since there is no code changes, close this support now.

**Comment by samveen** [comment]
@bybai I've figured out the solution today. My requirements were as below:
Every node has 2 IPs. Each node should have 3 entries in DNS: `node5`, `node5-priv`(alias of `node5`), and `node5-pub`(secondary IP).

Implementation for `xCAT-2.9.1`: 
We used a combination of the following attributes:
* `mac.mac` set to contain nodename-suffix: `"01:02:03:04:05:0E!node5-priv|01:02:03:05:0F!node6-pub"` *  `nics.nichostnamesuffixes` value of  "eth1!-pub,eth0!-priv" 
* empty value of `nicaliases.eth0`

This was required in `xCAT 2.9.1` as this version creates an alias `spare31-eth0` if `nichostnamesuffixes.eth0` is empty, which we wanted to avoid. Also if `nicaliases.eth0` was set, duplicate entries for `node5-priv` were added to `/etc/hosts`

Implementation in `xCAT-2.13.2`, which fixes a lot of the issues above:
* `mac.mac` can be bare: `01:02:03:04:05:0E|01:02:03:05:0F`
* `nics.nichostnamesuffixes` should contain `eth1!-pub` (bug with unnecesseary alias fixed). 
* `nics.nicaliases` should contain `eth0!node5-priv`
